src/Servers/Kestrel/Core/test/TlsListenerMiddlewareTests.cs

```diff
@@ -0,0 +1,569 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Net;
+using System.Net.Security;
+using System.Security.Authentication;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.InternalTesting;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Middleware;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Tests.TestHelpers;
+using Microsoft.AspNetCore.Server.Kestrel.Https;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Moq;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests;
+
+public partial class TlsListenerMiddlewareTests
+{
+    [Theory]
+    [MemberData(nameof(ValidClientHelloData))]
+    public Task OnTlsClientHelloAsync_ValidData(int id, byte[] packetBytes, bool nextMiddlewareInvoked)
+        => RunTlsClientHelloCallbackTest(id, packetBytes, nextMiddlewareInvoked, tlsClientHelloCallbackExpected: true);
+
+    [Theory]
+    [MemberData(nameof(InvalidClientHelloData))]
+    public Task OnTlsClientHelloAsync_InvalidData(int id, byte[] packetBytes, bool nextMiddlewareInvoked)
+        => RunTlsClientHelloCallbackTest(id, packetBytes, nextMiddlewareInvoked, tlsClientHelloCallbackExpected: false);
+
+    [Theory]
+    [MemberData(nameof(ValidClientHelloData_Segmented))]
+    public Task OnTlsClientHelloAsync_ValidData_MultipleSegments(int id, List<byte[]> packets, bool nextMiddlewareInvoked)
+        => RunTlsClientHelloCallbackTest_WithMultipleSegments(id, packets, nextMiddlewareInvoked, tlsClientHelloCallbackExpected: true);
+
+    [Theory]
+    [MemberData(nameof(InvalidClientHelloData_Segmented))]
+    public Task OnTlsClientHelloAsync_InvalidData_MultipleSegments(int id, List<byte[]> packets, bool nextMiddlewareInvoked)
+        => RunTlsClientHelloCallbackTest_WithMultipleSegments(id, packets, nextMiddlewareInvoked, tlsClientHelloCallbackExpected: false);
+
+    [Fact]
+    public async Task RunTlsClientHelloCallbackTest_DeterministinglyReads()
```

Comment:
Seems like you're trying to do the same thing as `RunTlsClientHelloCallbackTest_WithMultipleSegments` in this test, any reason this one should exist?

Reply:
I created `DeterministinglyReads` to also count how many times `ReadAsync` method should be invoked. I can incorporate that into `RunTlsClientHelloCallbackTest_WithMultipleSegments `, but I would need to calculate the range for each test input